cpp/src/arrow/compute/kernels/scalar_arithmetic.cc

```diff
@@ -684,8 +687,10 @@ void AddDecimalBinaryKernels(const std::string& name, ScalarFunction* func) {
   auto in_type256 = InputType(Type::DECIMAL256);
   auto exec128 = ScalarBinaryNotNullEqualTypes<Decimal128Type, Decimal128Type, Op>::Exec;
   auto exec256 = ScalarBinaryNotNullEqualTypes<Decimal256Type, Decimal256Type, Op>::Exec;
-  DCHECK_OK(func->AddKernel({in_type128, in_type128}, out_type, exec128));
-  DCHECK_OK(func->AddKernel({in_type256, in_type256}, out_type, exec256));
+  DCHECK_OK(func->AddKernel({in_type128, in_type128}, out_type, exec128, /*init=*/nullptr,
+                            constraint));
+  DCHECK_OK(func->AddKernel({in_type256, in_type256}, out_type, exec256, /*init=*/nullptr,
+                            constraint));
```

Comment:
Side note: do we have an issue open to extend this to Decimal32 and Decimal64, or are these kernels already registered elsewhere?